### DISCUSSION TITLE
BinaryFileResponse fails on Azure when running under Nginx on Linux (but works correctly on Windows/IIS)

### DISCUSSION THREAD

**Discussion by aokrasavin** [symfony/symfony#47059] [Q&A]
### Symfony version(s) affected

5.4, 6.0 and 6.1 confirmed, presumably others

### Description

A fairly simple controller which picks form input, chooses image from the disk and then returns it to webbrowser. The code works absolutely perfectly on local development instance of Laravel (Windows 10) or Azure Windows-based PHP service, but corrupts data when deployed on Azure PHP App Service  on Linux (under nginx).

1. All headers are delivered correctly with code 200
2. The binary payload for actual image has first 117 bytes stripped out. Eg after the headers, the binary data starts at byte 118, where byte 0 should have been. The rest of the bytes arrive correctly and at the end there are 117 empty bytes (filled with 0x00), so the total size is correct.
3. The issue exists for any files (not just images), for different types of images, for different security (eg HTTP or HTTPS)
6. The amount of bytes missing at start seems to be always the same and unrelated to image content (eg no magic byte sequence)

### How to reproduce

REPRODUCER

 1. Any recent version of Symfony (confirmed for versions 5.4, 6.0 and 6.1).
 2. Any recent PHP version (7, 8, etc)
 3. The problem manifests itself when run under nginx server on Azure PHP app service on Linux. To be clear - as of few months ago, PHP app stack on Azure is ONLY available on Linux and nginx. If you want to use PHP 8 or later - you MUST switch to this service. Original Windows/IIS based service will not longer be developed and is stuck at PHP 7.
 4. There is anecdotal evidence that the problem will manifest itself on any instance on nginx running under Linux nginx versions 1.20, 1.22 or 1.23

```
public function test()
{
$path = "......"; // a valid locally accessible path to a binary file (image, etc) of 30K or bigger
return new BinaryFileResponse($path);
}
```

### Possible Solution

The following fix works for me, but I do not fully understand the reason for the problem or possible implications, so would prefer someone with better in-depth knowledge to review prior to submitting pull request. The issue is in the following line:

```
namespace Symfony\Component\HttpFoundation;
......
class BinaryFileResponse extends Response
.....
    public function sendContent(): static
{
......
   stream_copy_to_stream($file, $out, $read);
......
}
```
to fix, I implemented the following - eg PHP native function stream_copy_to_stream is replaced with a custom function, which does (as far as I could tell) the same thing.
```
  function pipe_streams($in, $out)
    {
        $size = 0;
        while (!feof($in)) $size += fwrite($out,fread($in,8192));
        return $size;
    }

    public function sendContent(): static
{
......
   $this->pipe_streams($file, $out);
......
}
```

As people observed, this simple code does not support Ranges (so you need to disable it in headers). The buffer size is chosen randomly - we could have a lively argument of what it should be, but it does not materially changes the fix.

### Additional Context

I attached two versions of the same image returned by the SAME code running under Windows IIS (correct) and Linux (nginx) (corrupted
![172258214-12490c51-6769-43b7-a2ae-ca0b507bea2b](https://user-images.githubusercontent.com/54442217/180430982-34a03ac0-6f5b-472c-b35c-eeebb80bb620.png)
<img width="80" alt="172258194-9c888fb4-6543-4648-b097-6ceb5fce4fa8" src="https://user-images.githubusercontent.com/54442217/180430984-005f98f4-3bfc-47da-a540-9e15b05d8091.png">
)

In Laravel, the original code was

```
    $img = .... <get data object describing image>
    $response = response()->file(Storage::disk('public')->path($img->filename));
    $response->headers->set('Content-Type', $img->mime);
    $response->headers->set("Cache-Control", "max-age=86400");
    return $response;
```
 response()->file() and response->download() produce the same result - this is understandable as they only differ in headers

**Comment by michaelLoeffelmann** [comment]
i can confirm this behavior. i have exactly the same issue.

**Comment by MatTheCat** [comment]
The code does not look like Symfony :thinking: 

Anyways did you check if the request to the image includes a `Range` header?

**Reply by aokrasavin** [reply]
That is correct - the code is written in Laravel 9 which uses Symfony for all low-level data lifting. Ticket was raised with Laravel earlier (https://github.com/laravel/framework/issues/42685), who confirmed that "This is a Symfony issue.Please don't upgrade to Symfony v6.1 for now and wait until they release a patch release."
The fix I quoted works (eg not using  stream_copy_to_stream($file, $out, $read)). Right now I am hotpatching Symphony on my instance to make it work, but this is obvious not a correct solution. I played with "Range" - it has no effect on the outcome.

**Reply by aokrasavin** [reply]
I should also observe that similar problem is encountered by Drupal https://www.drupal.org/project/drupal/issues/3179014.

**Reply by MatTheCat** [reply]
The drupal issue seems different.

Does the problem still occur when upgrading symfony/http-foundation to the latest version?

If yes could you provide a reproducer?

**Reply by aokrasavin** [reply]
REPRODUCER

1. Any recent version of Symfony (confirmed for versions 5.4, 6.0 and 6.1). 
2. Any recent PHP version (7, 8, etc)
3. The problem manifests itself when run under nginx server on Azure PHP app service on Linux. To be clear - as of few months ago, PHP app stack on Azure is ONLY available on Linux and nginx. If you want to use PHP 8 or later - you MUST switch to this service. Original Windows/IIS based service will not longer be developed and is stuck at PHP 7. 
4. There is anecdotal evidence that the problem will manifest itself on any instance on nginx running under Linux nginx versions 1.20, 1.22 or 1.23

public function test()
    {
       $path = "......"; // a valid locally accessible path to a binary file (image, etc) of 30K or bigger
       return new BinaryFileResponse($path);
  }

**Reply by MatTheCat** [reply]
Please provide an application’s repository containing the minimum code required to reproduce your problem and running with Docker.

**Reply by aokrasavin** [reply]
I apologize - I do not know how do it while demonstrating stated manifestation conditions - I narrowed it down to calling a single function in any context - in effect as it currently stands, Symfony's BinaryFileResponse is completely broken under nginx/Linux. Revieweing the code - the implementation is inherently flawed and will break randomly on other platforms. I am not using Symfony directly - it is a second or third level dependency for me. Ultimately, I am happy with my fix - I provided the ticket as a courtesy to help other developers (I spent close to a day figuring out this one and it is frustratingly obscure in its manifestation). As far as my explanation of the fix stays on top of Google searches, I am content. and do not  particularly care if it is fixed ot not - so I am signing off of this discussion.
As an observation, the built-in method in question is used incorrectly in current Symfony code. It introduces a thread-racing condition which exposes itself in certain envrionments. With developers being forced to Nginx I expect more people will be encountering it.

**Reply by MatTheCat** [reply]
I used `BinaryFileResponse` with nginx countless time without any issue.

The “fix” you propose is impractical on top on breaking range requests so I don’t advise anybody to apply it.

Also by checking the laravel issue you failed to link I found https://github.com/laravel/framework/issues/42681#issuecomment-1154699658 As I don’t know which version you meant by `6.1` it may be a solution.

**Reply by aokrasavin** [reply]
The bug consistently manifests itself under nginx running on Linux on the cloud. I could not reproduce it on Windows version. I suspect that the reason for that is a different threading model - Linux is a "true" multithreaded system, while Windows compilations tend to be lazy and only pretend to be one. Because of this, thread-racing condition on Windows are either rare or impossible - this is why thread-racing issues are so frustrating. Since my original ticket on Laravel desk, I purposefully tested it on different versions of Symfony and consistently get the same outcome.

Laravel ticket is here https://github.com/laravel/framework/issues/42685 - if people search for the same header - it comes as the second link.

https://github.com/laravel/framework/issues/42681 is a different issue - it is not applicable here.

**Reply by aokrasavin** [reply]
I agree that my simple fix does nto support Range - I had it disabled on my system. This is the reason I did not want to put PULL request in - I wanted people with better understanding on low-level Symfony operations to apply proper fix.

**Comment by javierreguillo** [comment]
Hi,

I have the same problem with Symfony 5.4 on Azure App Service product using php 8.0 + nginx
The same project works fine on AWS and on premise... I think it could be some problem with the headers

Could you help me?

**Reply by aokrasavin** [reply]
While people are arguing who is responsible for the bug, you could apply my hotpatch (but you will have to reapply it every time you update). Locate BinaryFileResponse.php, method sendContent()

replace "stream_copy_to_stream(...);" with "$this->pipe_streams($file, $out);"

insert following utility method somewhere in the class

function pipe_streams($in, $out)
{
$size = 0;
while (!feof($in)) $size += fwrite($out,fread($in,8192));
return $size;
}

El presto - everything is working. This is also far better from memory consumption point of view - stream_copy_to_stream(...) is a well known memory hog. It is designed for copying LARGE streams  - it is pretty inefficient for small and medium size files (which is what Symfony normally deals with)

**Reply by MatTheCat** [reply]
@aokrasavin it would be really interesting if you could back your claims with some source :eyes:

**Reply by aokrasavin** [reply]
I am old-school developer, who likes to have a copy of original sources of everything to cross-check in case of any issues. This includes core PHP distribution.

In this particular case, the code of interest relate to function php_stream_copy_to_stream_ex (which could be found in streams.c). This is the method which does the actual data lifting - PHP_FUNCTION(stream_copy_to_stream) is purely a wrapper. 

The way it is written is seemingly to minimize number of individual  stream operations and maximize usage of memory. In particular, allocation of memory buffers is very brute-force (see line 1564 in 8.1.8 for example) unless they are very explicitly configured otherwise (which Symfony does not do, it relies on PHP global settings)

This is understandable - this method was written for moving files on local disk.

This is different from the way it is used inside BinaryFileResponse: 

1. BinaryFileResponse is (generally speaking) designed to serve content of files sitting on local storage to external web browser. Under majority of usage scenarios,  the speed limiting factor is going to be the web side of things - local storage will almost universally be faster (by order of magnitude, easily).
2. Instead of stuffing a huge amount of data into memory and trying to feed it to external pipe, it makes more sense to use much smaller local buffer. The current code will try to grab as much memory as possible up to PHP_STREAM_MMAP_MAX, which tends to be OS limit.
3. Note interesting assumptions made when making call to php_stream_mmap_possible() (line 1560)

**Reply by javierreguillo** [reply]
I think it's an azure error when you use linux + nginx + php 8 combined.
I don't know if changing symfony/http-foundation version would fix anything either

**Reply by aokrasavin** [reply]
Maybe. But considering that this combo (linux + nginx + php 8) becomes the facto standard and the ONLY deployment option on azure going forward (unless people happy to stay on PHP 7), it is something Symfony should be concerned about. As the things stand, any system relying on Symfony will have this part of functionality broken on Azure. 
There is anecdotal evidence that the problem affects non-Azure development setups - the key issue is the combo (linux + nginx + php 8).
From my point of view, the problem was more serious than images - I used them for neutral safe illustration of the problem. I am using my system to distribute firmware updates to IoT devices (OAT, etc). This REALLY messed things up.

**Reply by MatTheCat** [reply]
I just tested alpine/nginx/PHP 8 under Docker and `BinaryFileResponse` worked perfectly. If your issue is relative to something specific to your environment and you cannot provide a way for anyone to reproduce then it’ll be difficult to be concerned.

Also I did not understand anything from [your previous comment](https://github.com/symfony/symfony/discussions/47059#discussioncomment-3259935); if your concerns about `stream_copy_to_stream` only stem from your own understanding of PHP source code I’m afraid I won’t be able to follow :grimacing: 

However it sounds like you may want to use [`StreamedResponse`](https://symfony.com/doc/current/components/http_foundation.html#streaming-a-response) instead of `BinaryFileResponse`.

**Reply by aokrasavin** [reply]
My environment is vanilla Azure managed PHP app stack. Quite literally - out of the box. Microsoft manages the stack to PHP level. As you might have observed, other people experiencing it in identical circumstances. Microsoft certainly not using Alpine - my understanding is thet they are using own image.
The discussions about stream_copy_to_stream issues are well known in development community. Indeed, myself being proficient in C++ and C allows me to review and read framework code (to assembler level, if nessesary) - I also did some profiling (out of interest). I also have the benefit (courtesy of my company) of having access to various sophisticated and expensive profiling tools and run those routinely against development code as a matter of standard development procedures.
I will try StreamedResponse - unfortunately it means that I will have to ditch the convenience of Laravel 9 and do some customer code which I generally do not like. 
One way to reproduce is to get a free account on Azure (they are pretty liberal), get PHP managed app instance and run. The problem will be reproducable reliably. I am doing a lot of work with Azure on different projects and have direct access to their dev team, so enhoying some perks.

**Reply by javierreguillo** [reply]
Hello @aokrasavin, have you created a ticket in Azure support?

**Reply by aokrasavin** [reply]
I discussed it with the Azure support (and actually technical team). General conclusion was that it seems to be a bug in highly specific third-party piece of software, which has nothing to do with Azure. The gist was "Azure is not obligated to customize its infrastructure to every peculiarity of a third party software". They confirmed that there are no issues reported aroudn the same piece of functionality by other PHP-baserd deployments - it seems to be specific to Symfony and tools using it. They will accept a ticket if it could be demonstrated to be an Azure fault.

**Reply by javierreguillo** [reply]
It is a test of stream_copy_to_stream function on azure server
If this helps you, you could send it to them


`
    
    $srcFile = './downloads/apple-touch-icon-180x180.png';
    
    $src = fopen($srcFile, 'r');
    $out = fopen('php://output', 'w');
    
    stream_copy_to_stream($src, $out);
    
    fclose($src);
    fclose($out);

`
`

    $srcFile = './downloads/apple-touch-icon-180x180.png';

    $src = fopen($srcFile, 'r');
    $out = fopen('php://output', 'w');

    $size = 0;
    while (!feof($src)) {
        $size += fwrite($out, fread($src, 8192));
    }

    fclose($src);
    fclose($out);
`

As you can see, some bytes are removed (I highlighted them). 

![image](https://user-images.githubusercontent.com/20538884/182589789-3a5df6b4-b3fe-4ed0-8fc9-0265eefe8be4.png)

**Reply by MatTheCat** [reply]
[This thread](https://www.dolibarr.org/forum/t/installation-on-azure-blank-page-for-generated-pdfs/22030) seems eerily similar: using `readfile` fails but emulating it with `fread` works. Looks like Azure has an issue when you don’t chunk data you send to PHP output buffer :thinking:

**Reply by aokrasavin** [reply]
Very intetresting - yes - this is the kind of ammunition to argue the case!

**Reply by klodoma** [reply]
I am having the same issue see [here](https://stackoverflow.com/questions/73280423/php-readfile-content-corruption-on-azure-web-applinux)

I did some testing and it seems that all these `php` functions are affected:

`stream_copy_to_stream`
`readfile`
`fpassthru`

@aokrasavin: the BUG is coming out from symfony, laravel and wherever these functions are used. It's definitely not a library bug. It is a  native Azure Linux Web App bug, a very nasty one imo. I can confirm this on 2 Azure Installations. Considering the little feedback, it seems that PHP + Azure is not really used so much.

**Reply by MatTheCat** [reply]
Seems like the response body size has to meet a threshold greater than 8192 bytes for the problem to manifest (that may explain why [VenkateshDodda-MSFT does not reproduce](https://docs.microsoft.com/en-us/answers/questions/953763/php-stream-copy-to-stream-to-php-output-corrupts-d.html?childToView=955151#comment-955151)). Could you try increasing the chunks size until it appears again?

**Reply by MatTheCat** [reply]
@aokrasavin I think you can close this discussion as Microsoft answered.

**Comment by ghost** [comment]
While I can't drop a lot of details as the code is proprietary, we've seen this issue manifest in an existing codebase when jumping from Ubuntu 18.04 to 20.04 as part of an EOL upgrade.

Symfony version 5.x running php 7.4 under a debian variant container. The container itself did not change in any meaningful way, the host running the container workloads was what was upgraded.

Same situation, pulling a file to be served by a symfony server (no nginx) but we only got 8192 of the file (cutting off the top in our case).
